### ISSUE TITLE
Toggling the resource "pending" warning to improve library UX

### ISSUE THREAD

**Issue by richardliaw** [ray-project/ray#14051]
### Problem

It's often easier to 'fire and forget' for starting actors when resources are not available (and waiting for Ray to autoscale). However, we often see this nefarious, notorious error message:

```python
2020-01-13 18:15:09,447	WARNING worker.py:1062 -- The actor or task with ID ffffffffffffffff45b95b1c0100 is pending and cannot currently be scheduled. It requires {CPU: 1.000000} for execution and {CPU: 1.000000} for placement, but this node only has remaining {node:172.22.225.108: 1.000000}, {CPU: 2.000000}, {memory: 16.113281 GiB}, {GPU: 1.000000}, {object_store_memory: 5.566406 GiB}. In total there are 0 pending tasks and 2 pending actors on this node. This is likely due to all cluster resources being claimed by actors. To resolve the issue, consider creating fewer actors or increase the resources available to this Ray cluster. You can ignore this message if this Ray cluster is expected to auto-scale.
```

You might say "ok well you can solve that with placement groups". Unfortunately, this introduces complexity in various ways:

### User ends up needing to manage the creation of the placement group

Ideally, you have this:

```python
actors = []
# start N actors
for i in range(actors):
    new_pg.start()
    actor.options(pg).remote() 
    actors += [actor]
```
But in order to avoid the warning message: 

```python
# start N actors
for i in range(actors):
    groups += [(new_pg.start(), new_pg.ready())]
    
while pg in ray.wait(groups):
    actor.options(pg).remote()
    actors += [actor]
```

### If you want to reuse the placement group, you have to introduce a new state in your application level scheduling

Ideal:
```python
# PG reuse
for old_actor, _ in ray.wait(actors):
	old_actor.stop.remote()
    pg = old_actor.get_pg()
    actors += [new_actor.options(pg).remote()]
```

But in order to avoid the warning message: 

```python
# PG reuse
stopping_actors = []
while True:
	actor, _ = ray.wait(actors, timeout=0.1)
    if actor:
        stopping_actors.append(actor.stop.remote())
    if stopping_actors:
        old_actor, _ = ray.wait(stopping_actors, timeout=0.1)
        if old_actor:
            pg = old_actor.get_pg()
            actors += [new_actor.options(pg).remote()]
```

## Proposed Solution

Can we just have a configurable warning option?

```python
Actor.options(warn_about_placement_delay_s=60).remote
```

cc @ericl @edoakes @wuisawesome @krfricke

**Comment by ericl** [comment]
We can alternatively remove the error message if the cluster is
autoscaling, would that fix it? The autoscaler knows when tasks are truly
infeasible vs just waiting for scale up.

**Comment by wuisawesome** [comment]
Yeah, that sounds good. Create a flag to suppress that warning, then have the autoscaler set the flag. Then the autoscaler logs to driver when a task is truly infeasible. 

What happens if the user tries to spawn more long lived actors than `max_workers` though?

**Comment by richardliaw** [comment]
If you only set it for the autoscaling situation, you still need to separately address the `placement group reuse` use case.

**Comment by ericl** [comment]
We could also suppress the warning entirely for placement groups (just assume users know what they're doing in that case).

**Comment by ericl** [comment]
> What happens if the user tries to spawn more long lived actors than max_workers though?

Could the autoscaler also detect this (distinguishing between task and actor demands)?

**Comment by wuisawesome** [comment]
At the moment, no. But #13648  opens up some interesting possibilities.

**Comment by richardliaw** [comment]
Will it be possible to suppress the warning entirely then? That would be preferable imo.

**Comment by wuisawesome** [comment]
To handle the task vs actor case will require some reworking on the autoscaler side cc @AmeerHajAli

**Comment by rkooo567** [comment]
Why don't we just print this if the task is infeasible more than X seconds in a short term?

**Comment by wuisawesome** [comment]
alternatively, in the short term the autoscaler can print the message for both tasks and actors.

**Comment by richardliaw** [comment]
> Why don't we just print this if the task is infeasible more than X seconds in a short term?

X seconds is arbitrary, sometimes I need more, sometimes I need less

**Comment by krfricke** [comment]
Does "truly infeasible" mean that resources won't be available even when autoscaling?

In the current PG implementation for Tune, we have a rather high limit for scheduled placement groups (1000 at the moment). The rationale here is that we don't know how many actors are possible, so we try to request as many as we need. Those PGs that end up ready are then used for actors.

In a run of 1000 trials, if only 10 PGs can be started (e.g. via autoscaling) but Tune requests 1000 PGs, we would either need to suppress the warning for the rest of the 990 PGs, or we would need to detect that placing a placement group is infeasible and then stop requesting more groups.

Maybe something like `pg = placement_group(..., raise_if_infeasible=True)` which wouldn't create a pending PG but raise an error we could catch. Actually, even for normal actors/tasks it would be good to have something like this.

**Comment by richardliaw** [comment]
OK so maybe we can just have some flag that disables this warning? cc @wuisawesome

**Comment by wuisawesome** [comment]
To be clear, we're talking about disabling it in all cases (tasks and actors) right? I think we can do that relatively easily.

**Comment by richardliaw** [comment]
@wuisawesome that sounds good, maybe push a draft PR and we can take a look?

**Comment by wuisawesome** [comment]
Oh doing this per task is technically a core api change right?

**Comment by richardliaw** [comment]
There are numerous solutions in this thread that satisfy my application requirements:

1. Suppress the warning entirely for placement groups (just assume users know what they're doing in that case).
2. Configure the warning to be raised if the task is infeasible more than X seconds
3. Make the autoscaler print the warning instead of `worker.py`
4. Having some RayConfig that disables the warning upon task and actors.

Whether or not there's a 'core api change' is up to whichever solution we decide on. Personally, I like 1.

**Comment by edoakes** [comment]
+1 that all of @richardliaw's summarized options would also address the issue for Serve, and I'm also partial to solution (1).

**Comment by ericl** [comment]
> Suppress the warning entirely for placement groups (just assume users know what they're doing in that case).

Sounds good, and pretty easy (just ignore if there is a placement group resource).

**Comment by wuisawesome** [comment]
Can we do 1 and 5? 5 maybe not be necessary for tune/serve, but is nice because this warning is currently super annoying for task based workloads.

**Comment by wuisawesome** [comment]
Actually, maybe the more precise way to supress tasks could be to only display the warning for actor creation tasks (which fully satisfies 1. still). Thoughts?

**Comment by edoakes** [comment]
@wuisawesome Could you clarify how this would satisfy (1)? The warnings I see are for the actor creation task that requires the placement group resource.

**Comment by wuisawesome** [comment]
Ok, let me try this again: only show the warning for actor creation tasks that are not part of placement groups.

**Comment by pcmoritz** [comment]
Won't going with 1 mean, that this would be confusing for people who don't have an autoscaling cluster (or set the limits too low) and then their workload just gets stuck? I like doing 1 if the resources will eventually be fulfilled by the autoscaler, and giving a warning otherwise.

**Comment by ericl** [comment]
> Won't going with 1 mean that this would be confusing for people who don't have an autoscaling 

IIUC this only would apply to placement group users, which is a pretty constrained subset of the audience (mostly library developers).

So IMO, there is the short term solution and long term solution:
1. short term, narrowly disable the warning for placement group tasks and actors.
2. longer term, auto disable the warning if autoscaling is taking place (and maybe enable it again for placement groups)

**Comment by richardliaw** [comment]
@pcmoritz does Eric's above explanation make sense? If so, could you please +1?

It'd be good to get closure on this thread before we drag it on forever.

**Comment by pcmoritz** [comment]
So here is the resolution:
- We remove the warning in the placement group case as proposed in solution (1)
- At the same time we add a warning to the placement group docs that users are responsible to make sure their placement groups get ready with `ready` or `wait`

**Comment by richardliaw** [comment]
@wuisawesome @ericl is there an ETA for this?

**Comment by wuisawesome** [comment]
I'll try to get it done next week. 

Also fyi @robertnishihara
